File: Java-Kotlin/src/main/kotlin/tree/Binary Tree Right Side View.kt
package tree

import java.util.*

fun rightSideView(root: TreeNode?): List<Int> {
    val ans = mutableListOf<Int>()
    if (root == null) return ans
    val queue: Queue<TreeNode> = LinkedList()
    queue.add(root)
    while (!queue.isEmpty()) {
        var front = queue.peek()
        ans.add(front.`val`)
        val size = queue.size
        for (i in size downTo 1) {
            front = queue.peek();
            if (front.right != null) queue.add(front.right);
            if (front.left != null) queue.add(front.left);
            queue.poll()
        }
    }
    return ans
}

fun main() {
    val treeNode = TreeNode(1)
    treeNode.left = TreeNode(2)
    treeNode.right = TreeNode(3)
    print(rightSideView(treeNode))
}

File: C++/Tree/Binary Tree Right Side View.cpp
#include <bits/stdc++.h>
#include "TreeNode.h"

using namespace std;

/**
 * https://leetcode.com/problems/binary-tree-right-side-view/
 */

class Solution {
public:
    vector<int> rightSideView(TreeNode *root) {
        vector<int> result;
        queue<TreeNode *> queue;
        if (root) queue.push(root);
        while (!queue.empty()) {
            result.push_back(queue.front()->val);
            for (int i = queue.size(); i > 0; i--) {
                auto cur = queue.front();
                queue.pop();
                if (cur->right) queue.push(cur->right);
                if (cur->left) queue.push(cur->left);
            }
        }
        return result;
    }
};
